### ISSUE TITLE
[Bug] Does not detect class as implementing AsyncIterable.

### ISSUE THREAD

**Issue by AlexCheadle** [palantir/tslint#3997]
### Bug Report

- __TSLint version__: 5.10.0
- __TypeScript version__: 2.8.3
- __Running TSLint via__: (pick one) CLI 

#### TypeScript code being linted

```ts
class Stream implements AsyncIterable<number> {
  [Symbol.asyncIterator](): AsyncIterator<number> {
      return null as any;
  }
}
const stream: AsyncIterable<number> = new Stream();
const streamTyped: Stream = new Stream();

async function test() {
  // Gives no linter error.
  for await (const feature of stream) {
    // Do something
  }
  // Gives a linter error, await-promise
  for await (const feature of streamTyped) {
    // Do something
  }
}
```

with `tslint.json` configuration:

```json
{
    "defaultSeverity": "error",
    "extends": [
        "tslint:recommended"
    ],
    "jsRules": {},
    "rules": {
        "await-promise": true
    },
    "rulesDirectory": []
}
```
with tsconfig.json: 
```json
{
  "compilerOptions": {
    "target": "es2017",                          
    "module": "commonjs",                     
    "lib": ["esnext.asynciterable", "es2017"],                          
    "strict": true,                    
    "esModuleInterop": true,       
  }
}
```

#### Actual behavior
ERROR: <PATH>/test.ts[15, 31]: Invalid 'for-await-of' of a non-AsyncIterable value.

#### Expected behavior
No linting error.

**Comment by ajafff** [comment]
That's because the rule only recognises `AsyncIterableIterator<T>` which is a mistake I made when implementing it back then. This should be `AsyncIterable<T>` instead / in addition.

But there's more:
* it should allow every value that implements the async iteration protocol
* it should allow `Iterable<PromiseLike<T>>`. This is the fun one to implement 😄 
* the part of the rule checking `await` should allow everything that implements the Promise protocol instead of relying on names

---

I implemented all of these cases for my own linter runtime [Fimbullinter](https://github.com/fimbullinter/wotan). It's called [await-only-promise](https://github.com/fimbullinter/wotan/blob/master/packages/mimir/docs/await-only-promise.md) and is basically the better version of TSLint's `await-promise`.

**Comment by giladgray** [comment]
@ajafff any chance of porting your improvements back to tslint?

**Comment by zixia** [comment]
I ran into this issue today.

Any workaround right now?

**Comment by felixfbecker** [comment]
Same problem with `NodeJS.ReadableStream`, which implements `[Symbol.asyncIterator](): AsyncIterableIterator<string | Buffer>` (without an explicit `implements` clause)

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by ali-habibzadeh** [comment]
```typescript
interface IDog {
  breed: string;
  age: number;
}

async function getDog(id: number): Promise<IDog> {
  const dogs: IDog[] = [
    { breed: "german shepherd", age: 2 },
    { breed: "labrador", age: 4 }
  ];
  return dogs[id] || undefined;
}

const dogIterable: AsyncIterable<IDog> = {
  [Symbol.asyncIterator](): AsyncIterator<IDog, undefined> {
    let i = 0;
    return {
      async next(): Promise<IteratorResult<IDog>> {
        const dog = await getDog(i);
        i++;
        return {
          done: dog === undefined,
          value: dog
        };
      }
    };
  }
};

async function showDogs(): Promise<void> {
  for await (const dog of dogIterable) {
    console.log(dog);
  }
}

showDogs().catch( e=> { throw new Error(e) });
```

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
